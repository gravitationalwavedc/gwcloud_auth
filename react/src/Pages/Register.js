import React from 'react'
import {Button, Form, Grid, Header, Image, List, Message, Segment} from "semantic-ui-react";
import Link from 'found/lib/Link';
import {commitMutation} from "relay-runtime";
import {harnessApi} from "../index";
import {graphql} from "graphql";
import * as Enumerable from "linq";

class Register extends React.Component {
    constructor() {
        super();

        this.state = {
            username: "",
            email: "",
            firstname: "",
            lastname: "",
            password1: "",
            password2: "",
            verify: false,
            errors: null
        };
    }

    validate() {
        return this.state.username === "" || this.state.email === "" || this.state.firstname === "" || this.state.lastname === "" || this.state.password1.length < 5 || this.state.password2.length < 5 || this.state.password1 !== this.state.password2
    }

    submit() {
        commitMutation(harnessApi.getEnvironment("auth"), {
            mutation: graphql`mutation RegisterMutation($input: RegisterInput!)
                {
                  register(input: $input) 
                  {
                    result {
                      result
                      errors {
                        field,
                        messages
                      }
                    }
                  }
                }`,
            variables: {
                input: {
                    username: this.state.username,
                    email: this.state.email,
                    firstName: this.state.firstname,
                    lastName: this.state.lastname,
                    password1: this.state.password1,
                    password2: this.state.password2
                }
            },
            onCompleted: (response, errors) => {
                if (response.register.result.result)
                    this.setState({
                        ...this.state,
                        verify: true
                    });
                else
                    this.setState({
                        ...this.state,
                        errors: Enumerable.from(response.register.result.errors)
                    })
            },
        });
    }

    render() {
        return this.state.verify ? this.renderVerify() : this.renderForm()
    }

    renderVerify() {
        return (
            <Grid textAlign='center' style={{height: '100vh'}} verticalAlign='middle'>
                <Grid.Column style={{maxWidth: 450}}>
                    <Header as='h2' color='teal' textAlign='center'>
                        <Image src='/logo.png'/> Check Your Email
                    </Header>
                    <Message success>
                        Registration submitted. Please check your email.
                    </Message>
                </Grid.Column>
            </Grid>
        )
    }

    renderForm() {
        return (
            <Grid textAlign='center' style={{height: '100vh'}} verticalAlign='middle'>
                <Grid.Column style={{maxWidth: 450}}>
                    <Header as='h2' color='teal' textAlign='center'>
                        <Image src='/logo.png'/> Register an Account
                    </Header>
                    <Message error>
                        Do you have a LIGO.org account? <Link to='/auth/ligo/' activeClassName="selected"
                                                              exact {...this.props}>Login via LIGO.org</Link>
                    </Message>
                    <Form size='large'>
                        <Segment stacked>
                            {this.state.errors ? this.state.errors.where(e => e.field === "username").select(e => (
                                    <List bulleted floated="left">
                                        {Enumerable.from(e.messages).select((e, i) => (
                                            <List.Item key={i}>{e}</List.Item>
                                        ))}
                                    </List>
                                )
                            ).toArray() : null}
                            <Form.Input fluid icon='user' iconPosition='left' placeholder='Username'
                                        value={this.state.username}
                                        error={this.state.errors && this.state.errors.any(e => e.field === "username")}
                                        onChange={(e, c) => this.setState({...this.state, username: c.value})}/>
                            {this.state.errors ? this.state.errors.where(e => e.field === "email").select(e => (
                                    <List bulleted floated="left">
                                        {Enumerable.from(e.messages).select((e, i) => (
                                            <List.Item key={i}>{e}</List.Item>
                                        ))}
                                    </List>
                                )
                            ).toArray() : null}
                            <Form.Input fluid icon='envelope' iconPosition='left' placeholder='E-mail address'
                                        value={this.state.email}
                                        error={this.state.errors && this.state.errors.any(e => e.field === "email")}
                                        onChange={(e, c) => this.setState({...this.state, email: c.value})}/>
                            {this.state.errors ? this.state.errors.where(e => e.field === "firstName").select(e => (
                                    <List bulleted floated="left">
                                        {Enumerable.from(e.messages).select((e, i) => (
                                            <List.Item key={i}>{e}</List.Item>
                                        ))}
                                    </List>
                                )
                            ).toArray() : null}
                            <Form.Input fluid icon='address card' iconPosition='left' placeholder='First name'
                                        value={this.state.firstname}
                                        error={this.state.errors && this.state.errors.any(e => e.field === "firstName")}
                                        onChange={(e, c) => this.setState({...this.state, firstname: c.value})}/>
                            {this.state.errors ? this.state.errors.where(e => e.field === "lastName").select(e => (
                                    <List bulleted floated="left">
                                        {Enumerable.from(e.messages).select((e, i) => (
                                            <List.Item key={i}>{e}</List.Item>
                                        ))}
                                    </List>
                                )
                            ).toArray() : null}
                            <Form.Input fluid icon='address card' iconPosition='left' placeholder='Last name'
                                        value={this.state.lastname}
                                        error={this.state.errors && this.state.errors.any(e => e.field === "lastName")}
                                        onChange={(e, c) => this.setState({...this.state, lastname: c.value})}/>
                            {this.state.errors ? this.state.errors.where(e => e.field === "password1").select(e => (
                                    <List bulleted floated="left">
                                        {Enumerable.from(e.messages).select((e, i) => (
                                            <List.Item key={i}>{e}</List.Item>
                                        ))}
                                    </List>
                                )
                            ).toArray() : null}
                            <Form.Input
                                fluid
                                icon='lock'
                                iconPosition='left'
                                placeholder='Password'
                                type='password'
                                value={this.state.password1}
                                onChange={(e, c) => this.setState({...this.state, password1: c.value})}
                                error={this.state.errors && this.state.errors.any(e => e.field === "password1")}
                            />
                            {this.state.errors ? this.state.errors.where(e => e.field === "password2").select(e => (
                                    <List bulleted floated="left">
                                        {Enumerable.from(e.messages).select((e, i) => (
                                            <List.Item key={i}>{e}</List.Item>
                                        ))}
                                    </List>
                                )
                            ).toArray() : null}
                            <Form.Input
                                fluid
                                icon='lock'
                                iconPosition='left'
                                placeholder='Confirm password'
                                type='password'
                                value={this.state.password2}
                                onChange={(e, c) => this.setState({...this.state, password2: c.value})}
                                error={this.state.errors && this.state.errors.any(e => e.field === "password2")}
                            />

                            <Button color='teal' fluid size='large' disabled={this.validate()}
                                    onClick={() => this.submit()}>
                                Register
                            </Button>
                        </Segment>
                    </Form>
                </Grid.Column>
            </Grid>
        )
    }
}

export default Register;