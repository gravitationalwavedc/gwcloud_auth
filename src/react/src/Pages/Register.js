import React from 'react'
import {Button, Form, Grid, Header, Image, List, Message, Segment} from "semantic-ui-react";
import {commitMutation} from "relay-runtime";
import {harnessApi} from "../index";
import {graphql} from "graphql";
import * as Enumerable from "linq";
import ReCAPTCHA from 'react-google-recaptcha';

const reCAPTCHARef = React.createRef()

class Register extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            username: "",
            email: "",
            firstname: "",
            lastname: "",
            password1: "",
            password2: "",
            verify: false,
            errors: null
        };
    }

    validate() {
        return this.state.username === "" || this.state.email === "" || this.state.firstname === "" || this.state.lastname === "" || this.state.password1.length < 5 || this.state.password2.length < 5 || this.state.password1 !== this.state.password2 || !this.state.captcha
    }

    handleCaptcha = (key) => {
        this.setState({
            captcha: true,
            captchaResponse: key
        })
    }

    submit() {
        commitMutation(harnessApi.getEnvironment("auth"), {
            mutation: graphql`mutation RegisterMutation($input: RegisterInput!)
                {
                  register(input: $input) 
                  {
                    result {
                      result
                      errors {
                        field,
                        messages
                      }
                    }
                  }
                }`,
            variables: {
                input: {
                    username: this.state.username,
                    email: this.state.email,
                    firstName: this.state.firstname,
                    lastName: this.state.lastname,
                    password1: this.state.password1,
                    password2: this.state.password2,
                    captcha: this.state.captchaResponse
                }
            },
            onCompleted: (response, errors) => {
                if (response.register.result.result)
                    this.setState({
                        ...this.state,
                        verify: true
                    });
                else
                    this.setState({
                        ...this.state,
                        errors: Enumerable.from(response.register.result.errors)
                    })
            },
        });
    }

    render() {
        return this.state.verify ? this.renderVerify() : this.renderForm()
    }

    renderVerify() {
        return (
            <Grid textAlign='center' style={{height: '100vh'}} verticalAlign='middle'>
                <Grid.Column style={{maxWidth: 450}}>
                    <Header as='h2' color='teal' textAlign='center'>
                        <Image src='/logo.png'/> Check Your Email
                    </Header>
                    <Message success>
                        Registration submitted. Please check your email.
                    </Message>
                </Grid.Column>
            </Grid>
        )
    }

    renderForm() {
        return (
            <Grid textAlign='center' style={{height: '100vh'}} verticalAlign='middle'>
                <Grid.Column style={{maxWidth: 450}}>
                    <Header as='h2' color='teal' textAlign='center'>
                        <Image src='/logo.png'/> Register an Account
                    </Header>
                    <Message error>
                        Do you have a LIGO.org account? <a href='/auth/ligo/'>Login via LIGO.org</a>
                    </Message>
                    <Form size='large'>
                        <Segment.Group stacked>
                            <FormInputWithErrors
                                field="username"
                                icon="user"
                                placeholder="Username"
                                errors={this.state.errors}
                                value={this.state.username}
                                onChange={(e,c) => this.setState({...this.state, username: c.value})}
                            />
                            <FormInputWithErrors
                                field="email"
                                icon="envelope"
                                placeholder="E-mail address"
                                errors={this.state.errors}
                                value={this.state.email}
                                onChange={(e,c) => this.setState({...this.state, email: c.value})}
                            />
                            <FormInputWithErrors
                                field="firstName"
                                icon="address card"
                                placeholder="First name"
                                errors={this.state.errors}
                                value={this.state.firstname}
                                onChange={(e,c) => this.setState({...this.state, firstname: c.value})}
                            />
                            <FormInputWithErrors
                                field="lastName"
                                icon="address card"
                                placeholder="Last name"
                                errors={this.state.errors}
                                value={this.state.lastname}
                                onChange={(e,c) => this.setState({...this.state, lastname: c.value})}
                            />
                            <FormInputWithErrors
                                field="password1"
                                icon="lock"
                                placeholder="Password"
                                type="password"
                                errors={this.state.errors}
                                value={this.state.password1}
                                onChange={(e,c) => this.setState({...this.state, password1: c.value})}
                            />
                            <FormInputWithErrors
                                field="password2"
                                icon="lock"
                                placeholder="Confirm password"
                                type="password"
                                errors={this.state.errors}
                                value={this.state.password2}
                                onChange={(e,c) => this.setState({...this.state, password2: c.value})}
                            />
                            <Segment>
                                <ReCAPTCHA
                                    onChange={this.handleCaptcha}
                                    ref={reCAPTCHARef}
                                    sitekey='6LeXbtgUAAAAAPZ30BNcghORZcsHzLFeRs3qvrcH'
                                    theme='light'
                                    style={{
                                        display: "flex",
                                        justifyContent: "center",
                                    }}
                                />
                            </Segment>

                            <Button color='teal' fluid size='large' disabled={this.validate()}
                                    onClick={() => this.submit()}>
                                Register
                            </Button>
                        </Segment.Group>
                    </Form>
                </Grid.Column>
            </Grid>
        )
    }
}

function FormInputWithErrors(props) {
    return (
        <Segment>
            {props.errors ? props.errors.where(e => e.field === props.field).select(e => (
                    <List bulleted floated="left">
                        {Enumerable.from(e.messages).select((e, i) => (
                            <List.Item key={i}>{e}</List.Item>
                        ))}
                    </List>
                )
            ).toArray() : null}
            <Form.Input fluid icon={props.icon} iconPosition='left' placeholder={props.placeholder}
                        value={props.value} type={props.type}
                        error={props.errors && props.errors.any(e => e.field === props.field)}
                        onChange={props.onChange}/>
        </Segment>
    );
}

export default Register;