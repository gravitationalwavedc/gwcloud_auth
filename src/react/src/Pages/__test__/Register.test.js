import { createMockEnvironment } from 'relay-test-utils';
import { setHarnessApi } from '../../index';
import React from 'react';
import { describe, expect, it, jest } from '@jest/globals';
import TestRenderer from 'react-test-renderer';
import Register from '../Register';

// Set the harnessApi
setHarnessApi({
    getEnvironment: () => createMockEnvironment()
});

describe('registerTests', () => {
    it('register page renders correctly', () => {
        expect.hasAssertions();

        const router = {
            push: jest.fn(),
            replace: jest.fn(),
            go: jest.fn(),
            createHref: jest.fn(),
            createLocation: jest.fn(),
            isActive: jest.fn(),
            matcher: {
                match: jest.fn(),
                getRoutes: jest.fn(),
                isActive: jest.fn(),
                format: jest.fn()
            },
            addTransitionHook: jest.fn(),
            addNavigationListener: jest.fn()
        };
        const renderer = TestRenderer.create(<Register router={router} match={{}}/>);
        expect(renderer).toMatchSnapshot();
    });
});