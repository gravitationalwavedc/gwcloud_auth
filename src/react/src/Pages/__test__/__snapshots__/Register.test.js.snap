// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerTests register page renders correctly 1`] = `
<div
  className="h-100 container-fluid"
>
  <div
    className="h-100 row"
  >
    <div
      className="text-center d-none d-lg-block col-lg-6"
      style={
        Object {
          "background": "url(ligo-livingston-blue.png) no-repeat center",
          "backgroundSize": "cover",
          "minHeight": "100px",
        }
      }
    >
      <h1
        className="login-logo"
        style={
          Object {
            "color": "#F0FAFF",
            "marginTop": "197px",
          }
        }
      >
        <img
          src="GWC-Green192-768x521.png"
          style={
            Object {
              "height": "4rem",
              "marginBottom": "8px",
            }
          }
        />
         GWCloud
      </h1>
    </div>
    <img
      className="d-none d-lg-block"
      src="GWDC_grey_full_571x150.png"
      style={
        Object {
          "height": "3rem",
          "left": "50%",
          "margin": "1rem 0 0 1rem",
          "position": "absolute",
        }
      }
    />
    <img
      className="d-lg-none d-xs-block"
      src="GWDC_grey_full_571x150.png"
      style={
        Object {
          "height": "3rem",
          "margin": "1rem 0 0 1rem",
          "position": "absolute",
        }
      }
    />
    <div
      className="align-self-center col"
      style={
        Object {
          "marginTop": "5rem",
        }
      }
    >
      <div
        className="justify-content-md-center row"
      >
        <div
          className="col-lg-8 col-md-10"
        >
          <h2
            className="mb-2"
          >
            Register
          </h2>
          
          <form
            className=""
          >
            <div
              className="form-group"
            >
              <label
                className="form-label"
                htmlFor="email"
              >
                Email
              </label>
              <input
                className="text-left form-control"
                id="email"
                onChange={[Function]}
                type="email"
                value=""
              />
              <div
                className="invalid-feedback"
              />
            </div>
            <div
              className="row"
            >
              <div
                className="col"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                    htmlFor="firstName"
                  >
                    First Name
                  </label>
                  <input
                    className="text-left form-control"
                    id="firstName"
                    onChange={[Function]}
                    value=""
                  />
                  <div
                    className="invalid-feedback"
                  />
                </div>
              </div>
              <div
                className="col"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                    htmlFor="lastName"
                  >
                    Last Name
                  </label>
                  <input
                    className="text-left form-control"
                    id="lastName"
                    onChange={[Function]}
                    value=""
                  />
                  <div
                    className="invalid-feedback"
                  />
                </div>
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                className="form-label"
                htmlFor="password"
              >
                Password
              </label>
              <div
                className="input-group"
              >
                <input
                  aria-describedby="passwordHelpBlock"
                  className="text-left form-control"
                  id="password"
                  label="Password"
                  onChange={[Function]}
                  style={
                    Object {
                      "borderRadius": "0.25rem",
                    }
                  }
                  type="password"
                  value=""
                />
                <div
                  className="input-group-append"
                >
                  <a
                    className="p-2"
                    onClick={[Function]}
                  >
                    <svg
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M13.875 18.825A10.05 10.05 0 0112 19c-4.478 0-8.268-2.943-9.543-7a9.97 9.97 0 011.563-3.029m5.858.908a3 3 0 114.243 4.243M9.878 9.878l4.242 4.242M9.88 9.88l-3.29-3.29m7.532 7.532l3.29 3.29M3 3l3.59 3.59m0 0A9.953 9.953 0 0112 5c4.478 0 8.268 2.943 9.543 7a10.025 10.025 0 01-4.132 5.411m0 0L21 21"
                        strokeLinecap="round"
                        strokeLinejoin="round"
                        strokeWidth="2"
                      />
                    </svg>
                  </a>
                </div>
              </div>
              <small
                className="form-text text-muted"
                id="passwordHelpBlock"
              >
                At least 8 characters, at least one letter, shouldn't include your name or email, and can't be too common.
              </small>
              <div
                className="invalid-feedback"
              />
            </div>
            <div
              style={
                Object {
                  "display": "flex",
                }
              }
            />
            <button
              className="mb-4 mt-4 btn btn-primary btn-block"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Create new account
            </button>
            <small
              className="form-text text-muted"
              style={
                Object {
                  "marginTop": "-1rem",
                }
              }
            >
              This site is protected by reCAPTCHA and the Google
              <a
                href="https://policies.google.com/privacy"
              >
                 Privacy Policy
              </a>
               and
              <a
                href="https://policies.google.com/terms"
              >
                 Terms of Service
              </a>
               apply.
            </small>
          </form>
          <p
            className="pb-4"
          >
            <a
              onClick={[Function]}
            >
              Already have an account?
            </a>
          </p>
        </div>
      </div>
      <div
        className="justify-content-md-center row"
        style={
          Object {
            "margin": "4rem 0 2rem 0",
          }
        }
      >
        <div
          className="col-lg-8 col-md-10"
        >
          <div
            className="align-items-end row"
          >
            <div
              className="col"
            >
              <img
                src="ncris_logo.png"
                style={
                  Object {
                    "maxHeight": "4rem",
                    "maxWidth": "100%",
                  }
                }
              />
            </div>
            <div
              className="text-center col"
            >
              <img
                src="ozgrav-logo-array-tk005-final-invert-textedit.png"
                style={
                  Object {
                    "maxHeight": "4.5rem",
                    "maxWidth": "100%",
                  }
                }
              />
            </div>
            <div
              className="text-right col"
            >
              <img
                src="aal-logo-rp-revised-wwording-800px.png"
                style={
                  Object {
                    "maxHeight": "3rem",
                    "maxWidth": "100%",
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-6 col-md-12 offset-lg-6"
    />
  </div>
</div>
`;
